### ISSUE TITLE
Clarify CONSTRUCT with blank nodes and duplicate solutions

### ISSUE THREAD

**Issue by cygri** [w3c/sparql-12#87]
**TL;DR**: The SPARQL spec is ambiguous about the behaviour of blank nodes in `CONSTRUCT` templates when there are duplicate solutions. The spec should be clarified to resolve the ambiguity. Most (but not all) implementations create different blank nodes for duplicate solutions.

---

The SPARQL spec [says about blank nodes in `CONSTRUCT` templates](https://www.w3.org/TR/2013/REC-sparql11-query-20130321/#templatesWithBNodes) (emphasis mine):

> The blank node labels are **scoped to the template for each solution**. If the same label occurs twice in a template, then there will be one blank node created for each query solution, but there will be **different blank nodes for** triples generated by **different query solutions**.

This does not make clear what should happen if the solution sequence for `CONSTRUCT` contains duplicates. Consider this query:
```
CONSTRUCT { _:s :p 1 }
WHERE { {} UNION {} }
```
Does the solution sequence of the `WHERE` clause contain two different solutions? Or does it contain a single solution, twice? The language of the spec leaves this ambiguous, and unfortunately the formalism in the spec only treats `SELECT` queries and thus doesn't help to disambiguate.

So it is unclear whether the query produces a one-triple graph or a two-triple graph.

## Implementation behaviour
Implementation | Behaviour
---|---
Blazegraph | One triple
Comunica | Blank nodes in `CONSTRUCT` template not supported
Jena | Two triples
librdf | Two triples (needs a slightly different query due to [this bug](https://github.com/RDFLib/rdflib/issues/910))
Rasqal | Two triples
RDF4J | Two triples
Virtuoso | Two triples (needs a slightly different query due to a bug with empty-branch `UNION`)

## Summary
Almost all implementations take the view that duplicates in the solution sequence count as separate solutions, and produce different blank nodes from the `CONSTRUCT` template.

The spec should be clarified to unambiguously reflect this interpretation.

One way to achieve this would be to extend [Chapter 18, *Definition of SPARQL*](https://www.w3.org/TR/2013/REC-sparql11-query-20130321/#sparqlDefinition) to cover `CONSTRUCT`.

**Comment by rubensworks** [comment]
For reference, Comunica now properly supports bnodes in CONSTRUCT queries.
It also produces [two distinct triples](http://query.linkeddatafragments.org/#transientDatasources=http%3A%2F%2Ffragments.dbpedia.org%2F2016-04%2Fen&query=PREFIX%20%3A%20%3Curn%3Adefault%3A%3E%0A%0ACONSTRUCT%20%7B%20_%3As%20%3Ap%201%20%7D%0AWHERE%20%7B%20%7B%7D%20UNION%20%7B%7D%20%7D).

**Comment by afs** [comment]
> The blank node labels are scoped to the template for each solution.

A solution is one row from the WHERE clause.

[18.1.8 Solution Mapping](https://www.w3.org/TR/sparql11-query/#sparqlSolutions)
> A solution mapping is a mapping from a set of variables to a set of RDF terms. We use the term 'solution' where it is clear.

`WHERE { {} UNION {} }` has two solutions (solution mappings) in the solution sequence.

The template is used twice.
Is there text in the spec that implies duplicates are considered in any special way?

> different blank nodes for triples generated by different query solutions.

The template is used twice so there are two different blank nodes, one from each template use.

If one system behaves differently, it would be better to take the matter up with that system first to give them a chance to respond.

**Comment by cygri** [comment]
@afs 

> The template is used twice so there are two different blank nodes

The spec doesn't say that there is a different blank node for each use of the template. It says: *“there will be different blank nodes for triples generated by different query solutions”.*

So the question, is *what makes two query solutions the same or different?* According to the spec text that you have helpfully quoted, a solution is a mapping from a set of variables to RDF terms. According to that definition, the only coherent answer has to be: *Two mappings are the same if they bind the same values to the same variables. Otherwise, they are different.* Is there any other coherent reading?

So let me repeat: “there will be different blank nodes for triples generated by different query solutions”. In the `{} UNION {}` example, the template is used twice, but *over the same solution* (the empty solution). *So there should be the same blank node generated in both cases.*

That is where one is taken by a precise reading of the spec.

**Comment by afs** [comment]
So what you propose is a text change from:  
"there will be different blank nodes for triples generated by different query solutions”.  
to 
"there will be different blank nodes for triples generated by each element in the list that is the solution sequence".
?

**Comment by cygri** [comment]
That would be one way to resolve the ambiguity regarding handling of duplicate solutions. Another way would be to add: “If the solution sequence contains duplicate solutions, the duplicates are considered as different solutions for the purpose of blank node generation.” Another way would be, as stated in the issue description, to extend the formalism in Chapter 18 to cover CONSTRUCT and not just SELECT.

**Comment by VladimirAlexiev** [comment]
Which alternative supports Streaming? I think duplicate elimination may not be scalable

**Comment by TallTed** [comment]
@cygri 
> **Virtuoso**	Two triples (needs a slightly different query due to a bug with empty-branch `UNION`)

Can you provide any additional detail on this, optimally as [a report on the Community Forum](http://community.openlinksw.com) or [an issue on the Open Source Project](https://github.com/openlink/virtuoso-opensource/issues)?  It would be especially helpful if such reports could include any known version specificity (primarily differentiating between Open Source v7, Commercial v7, Commercial v8, but more granular is almost always better if possible).
